controller/node_controller.go

```diff
@@ -995,6 +1001,139 @@ func (nc *NodeController) syncNodeStatus(pod *corev1.Pod, node *longhorn.Node) e
+	namespaces := []lhtypes.Namespace{lhtypes.NamespaceMnt, lhtypes.NamespaceNet}
+	nc.syncPackagesInstalled(kubeNode, node, namespaces)
+	nc.syncMultipathd(node, namespaces)
+	nc.syncNFSClientVersion(kubeNode, node, namespaces)
+}
+
+func (nc *NodeController) syncPackagesInstalled(kubeNode *corev1.Node, node *longhorn.Node, namespaces []lhtypes.Namespace) {
+	osImage := strings.ToLower(kubeNode.Status.NodeInfo.OSImage)
+	queryPackagesCmd := ""
+	packages := []string{}
+	switch {
+	case strings.Contains(osImage, "ubuntu"):
+	case strings.Contains(osImage, "debian"):
+		queryPackagesCmd = "dpkg -l | grep -w"
+		packages = []string{"nfs-common", "open-iscsi", "cryptsetup", "dmsetup"}
+	case strings.Contains(osImage, "centos"):
+	case strings.Contains(osImage, "fedora"):
+	case strings.Contains(osImage, "rocky"):
+	case strings.Contains(osImage, "ol"):
+		queryPackagesCmd = "rpm -q"
+		packages = []string{"nfs-utils", "iscsi-initiator-utils", "cryptsetup", "device-mapper"}
+	case strings.Contains(osImage, "suse"):
+		queryPackagesCmd = "rpm -q"
+		packages = []string{"nfs-client", "open-iscsi", "cryptsetup", "device-mapper"}
+	case strings.Contains(osImage, "arch"):
+		queryPackagesCmd = "pacman -Q"
+		packages = []string{"nfs-utils", "open-iscsi", "cryptsetup", "device-mapper"}
+	case strings.Contains(osImage, "gentoo"):
+		queryPackagesCmd = "qlist -I"
+		packages = []string{"net-fs/nfs-utils", "sys-block/open-iscsi", "sys-fs/cryptsetup", "sys-fs/lvm2"}
+	default:
+		node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeRequiredPackages, longhorn.ConditionStatusFalse,
+			string(longhorn.NodeConditionReasonUnknownOS),
+			fmt.Sprintf("Node %v has an unknown OS image %v, please make sure the necessary packages are installed", node.Name, osImage))
+		return
+	}
+
+	nsexec, err := lhns.NewNamespaceExecutor(lhtypes.ProcessNone, lhtypes.HostProcDirectory, namespaces)
+	if err != nil {
+		node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeRequiredPackages, longhorn.ConditionStatusFalse,
+			string(longhorn.NodeConditionReasonNamespaceExecutorErr),
+			fmt.Sprintf("Get namespace executor failed: %v", err))
+		return
+	}
+	var NotFoundPkgs []string
+	for _, pkg := range packages {
+		args := []string{pkg}
+		if _, err := nsexec.Execute(nil, queryPackagesCmd, args, lhtypes.ExecuteDefaultTimeout); err != nil {
+			nc.logger.WithError(err).Debugf("Package %v is not found in node %v", pkg, node.Name)
+			NotFoundPkgs = append(NotFoundPkgs, pkg)
+		}
+	}
+
+	if len(NotFoundPkgs) > 0 {
+		node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeRequiredPackages, longhorn.ConditionStatusFalse,
+			string(longhorn.NodeConditionReasonPackagesNotInstalled),
+			fmt.Sprintf("Node %v is missing the following packages: %v, please make sure they are installed", node.Name, NotFoundPkgs))
```

Comment:
Node name is unnecessary because the condition is under node.status.

```suggestion
			fmt.Sprintf("Missing packages: %v", NotFoundPkgs))
```

Reply:
Modified. Thanks